Add Microsoft.Extensions.ApiDesription.Server project and package

dougbu opened the pull request:
- #8242 2 of 2

dougbu committed:
Add Microsoft.Extensions.ApiDesription.Server project and package
- #8242 2 of 2

dougbu commented:
Also addresses #4912 because I didn't restore the `GetProjectReferenceMetadata` task

dougbu commented:
@natemcmaster I'd of course appreciate any input you have on the MSBuild changes

dougbu commented on src/Mvc/Extensions.ApiDescription.Server/src/build/Microsoft.Extensions.ApiDescription.Server.props:
```diff
@@ -0,0 +1,31 @@
+  <!--
+    Settings users may update as they see fit.
+  -->
+  <PropertyGroup>
```
@glennc do you have any input on the names of these user-visible properties?

glennc replied:
As we discussed can we replace Retrieve with Generate in these properties.

dougbu committed:
Add missing project reference provider

dougbu committed:
!fixup! `<Reference>` issues and PR comments
- remove incorrect `$(IsImplementationProject)` settings and re-run .\eng\scripts\GenerateProjectList.ps1
- address PR comments
  - correct extra word in a comment that @JunTaoLuo noticed

nits:
- minor formatting changes in Microsoft.Extensions.ApiDescription.Server.props
- remove the project file's extension from the file list cache filename

JunTaoLuo commented on eng/ProjectReferences.props:
```diff
@@ -8,7 +8,10 @@
-    <ProjectReferenceProvider Include="Microsoft.Extensions.ApiDescription.Tasks" ProjectPath="$(RepoRoot)src\Mvc\Extensions.ApiDescription.Client\src\Microsoft.Extensions.ApiDescription.Client.csproj"  />
+    <ProjectReferenceProvider Include="Microsoft.Extensions.ApiDescription.Client" ProjectPath="$(RepoRoot)src\Mvc\Extensions.ApiDescription.Client\src\Microsoft.Extensions.ApiDescription.Client.csproj"  />
+    <ProjectReferenceProvider Include="Microsoft.Extensions.ApiDescription.Server" ProjectPath="$(RepoRoot)src\Mvc\Extensions.ApiDescription.Server\src\Microsoft.Extensions.ApiDescription.Server.csproj"  />
+    <ProjectReferenceProvider Include="dotnet-getdocument" ProjectPath="$(RepoRoot)src\Mvc\dotnet-getdocument\src\dotnet-getdocument.csproj" />
```
Hmm I think you're still missing some pieces. 
```
F:\workspace\_work\1\s\eng\targets\ResolveReferences.targets(212,5): error MSB3245: Could not resolve this reference. Could not locate the package or project for "dotnet-getdocument". Did you update baselines and dependencies lists? See docs/ReferenceResolution.md for more details. [F:\workspace\_work\1\s\src\Mvc\Extensions.ApiDescription.Server\src\Microsoft.Extensions.ApiDescription.Server.csproj]

```

JunTaoLuo commented on src/Mvc/Extensions.ApiDescription.Server/src/Microsoft.Extensions.ApiDescription.Server.nuspec:
```diff
@@ -0,0 +1,27 @@
+    <file src="$artifactsBinDir$\dotnet-getdocument\$configuration$\netcoreapp2.1\publish\*.*" target="tools" />
+    <file src="$artifactsBinDir$\GetDocumentInsider\$configuration$\net461\GetDocument.Insider.*" target="tools\net461" />
+    <file src="$artifactsBinDir$\GetDocumentInsider\x86\$configuration$\net461\GetDocument.Insider.*" target="tools\net461-x86" />
+    <file src="$artifactsBinDir$\GetDocumentInsider\$configuration$\netcoreapp2.1\GetDocument.Insider.*" target="tools\netcoreapp2.1" />
```
This seems to be only compiled for netcoreapp2.0.

JunTaoLuo commented on src/Mvc/Extensions.ApiDescription.Server/src/build/Microsoft.Extensions.ApiDescription.Server.props:
```diff
@@ -0,0 +1,31 @@
+    <OpenApiRetrieveDocuments Condition=" '$(OpenApiRetrieveDocuments)' == '' ">true</OpenApiRetrieveDocuments>
+
+    <!--
+      If 'true' (the default if $(OpenApiRetrieveDocuments) is 'true'), will retrieve of Open API documents after every
```
```suggestion
      If 'true' (the default if $(OpenApiRetrieveDocuments) is 'true'), will retrieve Open API documents after every
```

dougbu committed:
!fixup! `GetDocumentCommand` was still using .NET Core App 2.0 define

dougbu commented:
Bit more overlap with #10667 but resolution won't be too difficult

JunTaoLuo commented:
/azp run

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Server/src/build/Microsoft.Extensions.ApiDescription.Server.props:
```diff
@@ -0,0 +1,32 @@
+      Options added to the Open API retrieval tool ('dotnet-getdocument') command line. Available options control
+      console output: 'no-color', 'prefix-output' and 'verbose'. All require a double-dash prefix.
+    -->
+    <OpenApiGetDocumentOptions Condition=" '$(OpenApiGetDocumentOptions)' == '' "></OpenApiGetDocumentOptions>
```
This line doesn't do anything. "If OpenApiGetDocumentOptions is empty, assign OpenApiGetDocumentOptions  to empty."

dougbu replied:
Good catch! This property should be `true` by default

dougbu replied:
Oh, wait. This is here only for documentation purposes. Should be empty by default.

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Server/src/build/Microsoft.Extensions.ApiDescription.Server.props:
```diff
@@ -0,0 +1,32 @@
+    <!--
+      If 'true' (the default), enable retrieval of Open API documents. Otherwise, this feature is completely disabled.
+    -->
+    <OpenApiRetrieveDocuments Condition=" '$(OpenApiRetrieveDocuments)' == '' ">true</OpenApiRetrieveDocuments>
```
Why would I install this package and then disable it's functionality completely? Seems like it would be simpler to say "if you want to disable the feature, uninstall the package"

dougbu replied:
Users are not usually aware of this package at all. We'll be working with NSwag and SwashBuckle to ship with a dependency on it in their existing NSwag.AspNetCore and SwashBuckle.AspNetCore. Those packages have a lot of features and some users may wish to disable this particular one.

That is, we decided on opt-out for saving Open API docs after build.

dougbu commented on src/Mvc/Extensions.ApiDescription.Server/src/Microsoft.Extensions.ApiDescription.Server.csproj:
```diff
@@ -0,0 +1,64 @@
+    <NoPackageAnalysis>true</NoPackageAnalysis>
+
+    <!-- Included primarily to ensure dotnet-getdocument and GetDocument.Insider can be referenced. -->
+    <TargetFrameworks>netcoreapp2.1;net461</TargetFrameworks>
```
See #10667. That adds `runtimeconfig.template.json` files for `dotnet-getdocument` and `GetDocument.Insider`

dougbu replied:
Oh, I also confirmed this works fine in a 3.0-only environment with 3.0 projects

dougbu committed:
!fixup! Close an element

dougbu commented:
@natemcmaster @JunTaoLuo builds are looking much better now 😺

dougbu committed:
Rename zee properties
- rename targets and the project capability for consistency
- left the `OpenApiGetDocuments` target alone because it's used from client projects and isn't about generation

rynowak commented on src/Mvc/Extensions.ApiDescription.Server/src/build/Microsoft.Extensions.ApiDescription.Server.targets:
```diff
@@ -0,0 +1,39 @@
+  </ItemGroup>
+
+  <Target Name="OpenApiGetDocuments" Returns="@(_OpenApiProjectDocuments)">
+    <ReadLinesFromFile File="$(_OpenApiDocumentsCache)">
```
Is this actually a cache? or is this just the output of generating the documents?

dougbu replied:
### TL;DR
Is your question aimed at my naming (and life) choices? I don't see much consistency in how similar MSBuild properties are named and chose something that fit.

Make a suggestion if you feel strongly please.

### Details
This file is many things&hellip;

It's a cache in the sense seen in lots of MSBuild code: It's a way to avoid redoing an action (calling a task, executing a too, whatever). In particular, this file avoids reopening the app whenever a client project calls `OpenApiGetDocuments`.

Yes, as an implementation detail, this file is _one of_ the files written as a result of running the tool. It lists the document files.

----
The file is also like many used in MSBuild code in that it summarizes multiple inputs or outputs of a task or tool. For example, I could do the `Inputs=$(TargetPath)` &hellip; `Outputs=$(_OpenApiDocumentsCache)` batching trick to avoid running `dotnet-getdocument` even if there are multiple document files.

----
Finally, this file is the communication mechanism between the `GenerateOpenApiDocuments` and `OpenApiGetDocuments` targets.

rynowak commented on src/Mvc/Extensions.ApiDescription.Server/src/build/Microsoft.Extensions.ApiDescription.Server.targets:
```diff
@@ -0,0 +1,39 @@
+
+    <Message Importance="high" Text="%0ARetrieveOpenApiDocuments:" />
+    <Message Importance="high" Text="  $(_Command)" />
+    <Exec Command="$(_Command)" />
```
What are the caching expectations of this? Should we be adding some things to `FileWrites` so they can be removed by incremental clean?

dougbu replied:
I have been thinking about `FileWrites` both here and in the client-side glue. But, every time I start making those changes, I hit the "What if the user plans to check these files in?" question. These files aren't always throw-away. But, I could probably complicate things enough to avoid that concern e.g. with conditions on `$(OpenApiDocumentsDirectory)` matching the default.

I'll add a note to #4923 reminding me to work through the implications and do _something_.

rynowak replied:
Discussed in person. Right now there is no input/output checking and this command will re-run every time you build, even if the assembly does not change.

This is tracked separately and will be improved in the next preview.

dougbu commented:
Fair number of comments from a fair number of people but I'm not sure whether anything is a blocking concern. @rynowak (for example), do you approve of leaving things as-is until I address the #4923 list?

dougbu committed:
!fixup! Avoid ending a quoted command-line argument with \"

dougbu committed:
Correct `OpenApiGetDocuments` target name in buildMultiTargeting/Microsoft.Extesions.ApiDescription.Client.targets

nits:
- more s/retrieve/generate/

dougbu commented:
:up::date: for smaller fixes and a rebase on 'master'. Could someone please approve?

dougbu committed:
Merge branch 'master' into dougbu/add.server-side.8242.2

dougbu commented:
Bump @rynowak or @pranavkm

dougbu commented:
Retargeted to preview 6 branch. PR validation begun again.

/cc @mkArtakMSFT

dougbu commented:
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=209867

dougbu commented:
Just another Helix hang unrelated to anything in this PR. Merging&hellip;

aspnet-hello commented:
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2521
